app/controllers/facility_user_reservations_controller.rb

```diff
@@ -18,35 +19,39 @@ def index
                      .paginate(page: params[:page])
   end
 
-  # PUT /facilities/:facility_id/users/:user_id/reservations/:id/cancel
+  # PUT /facilities/:facility_id/users/:user_id/reservations/:order_detail_id/cancel
   def cancel
-    order_detail = user_order_details.find(params[:id])
-    raise ActiveRecord::RecordNotFound unless order_detail.reservation.try(:can_cancel?)
-    order_detail.transaction do
-      if cancel_with_fee(order_detail)
-        flash[:notice] = text("cancel.success")
-      else
-        flash[:error] = text("cancel.error")
-        raise ActiveRecord::Rollback
-      end
+    reservation = @order_detail.reservation
```

Comment:
I refactored to make it work like:

* If the `OrderDetail` has no `Reservation`, 404
* If the `Reservation` is not canceled,
  - 404 if the `Reservation` can't be canceled
  - Attempt to cancel the `Reservation` and roll back if it fails
* If the `Reservation` is canceled, flash a success notice
* If the `Reservation` is not canceled, flash an error notice
* Redirect to the index